### ISSUE TITLE
feature: Allow more than one PostCreateHook when creating a control plane

### ISSUE THREAD

**Issue by MikeSpreitzer** [kubestellar/kubeflex#225]
### Feature Description

Currently a control plane construction can invoke at most one PostCreateHook. This feature would allow more than one.

This would be very helpful in untangling dependencies among various repos in the KubeStellar organization. Currently, wherever more than one thing has to be added to a control plane, something has to define one PostCreateHook that adds all of it. Consider, for example, a WDS: it needs both the kubestellar controller-manager from the kubestellar repo and will soon also need the transport controller from the ocm-transport-plugin repo.

### Proposed Solution

In `ControlPlaneSpec`, replace

```go
	PostCreateHook     *string           `json:"postCreateHook,omitempty"`
	PostCreateHookVars map[string]string `json:"postCreateHookVars,omitempty"`
```

with

```go
	PostCreateHooks     []PostCreateHookUse           `json:"postCreateHooks,omitempty"`
```

and define

```go
type PostCreateHookUse struct {
    HookName *string `json:"hookName"`
    Vars map[string]string `json:"vars,omitempty"`
}
```

In the CLI, allow multiple `-p --set ... --set` trains in series.

### Want to contribute?

- [ ] I would like to work on this issue.

**Comment by pdettori** [comment]
@MikeSpreitzer proposal looks good. It will have to be a non-backward compatible change but I guess it's fine. Why the name `PostCreateHookUse` and not just `PostCreateHook` ?

**Comment by francostellari** [comment]
If possible, can we make a decision #222 first?
These changes would probably require a significant rebase.
The required changes to the chart will also be included in this PR.

**Comment by pdettori** [comment]
@francostellari do you mean that you prefer we merge 222 first, right ?

**Comment by francostellari** [comment]
> @francostellari do you mean that you prefer we merge 222 first, right ?

Yes, if possible with the other priorities

**Comment by MikeSpreitzer** [comment]
@pdettori : OK, I was not sure about whether backward compatibility would be needed. If so then I would suggest the following roadmap. At first, instead of doing the _replacement_ that I sugggested, make it an _addition_. Both the old and new ways of requesting PCH usage are interpreted by KubeFlex. Then switch clients to using only the new way. Then we can retire the old way.

I make a distinction between a PostCreateHook, which is an API object defining something to do when a control plane is created, and a _usage_ of a PostCreateHook, which is a statement that a particular hook should be used with a particular control plane.

**Comment by MikeSpreitzer** [comment]
BTW, if we merged this before #222 then #222 would not have to add a dependency cycle to ks/ks packaging.md ; that is something I would really like to avoid.

**Comment by ezrasilvera** [comment]
@MikeSpreitzer It seems an interesting capability. We need to remember we still need to see how much we want to be coupled with KF regarding deployment. 
I'm not sure that deployment workloads is a common feature for space or CP managers.  This is a very strong coupling and need to be discussed.

**Comment by pdettori** [comment]
@ezrasilvera I thought we kind of invented space / cp managers, so we should have some freedom to define the capabilities ;). As for concerns about coupling, just remember that all that can be done with the pch can also be done with helm or other CLI tools (such as clusteradm or kubectl), so as long as we avoid the cyclic dependency issue (which I think now is mitigated by the addition of pch vars) we should be ok. I just see pch as a facility to improve/simplify UX when using KF.

**Comment by ezrasilvera** [comment]
> @ezrasilvera I thought we kind of invented space / cp managers, so we should have some freedom to define the capabilities ;). As for concerns about coupling, just remember that all that can be done with the pch can also be done with helm or other CLI tools (such as clusteradm or kubectl), so as long as we avoid the cyclic dependency issue (which I think now is mitigated by the addition of pch vars) we should be ok. I just see pch as a facility to improve/simplify UX when using KF.

@pdettori I agree, this is a nice differentiator in UX and user experience.  I just wanted to try to maintain also the "regular deployment path"  this is mainly a documentation point (i.e., document how we can deploy the KS components directly without PCH) . I'm sure you don't object to that .

**Comment by MikeSpreitzer** [comment]
@ezrasilvera : so I think the net of your comments above for this PR is zero? As long as we are going to have PCHs, we should not restrict a control plane to use only one.

**Comment by ezrasilvera** [comment]
> @ezrasilvera : so I think the net of your comments above for this PR is zero? As long as we are going to have PCHs, we should not restrict a control plane to use only one.

I'm fine with that. We don't promise any ordering between the PCHs , right ?  Not even that there is a blocking between each ..

**Comment by MikeSpreitzer** [comment]
@ezrasilvera : good point about ordering. Yes, I think we can start with no ordering. If we discover a need for ordering later then we can deal with it then.
